mapper/tedge_mapper/src/mapper.rs

```diff
@@ -86,36 +91,108 @@ impl Mapper {
-            match self.converter.convert(message.payload_str()?) {
-                Ok(mapped) => {
-                    self.client
-                        .publish(mqtt_client::Message::new(&self.config.out_topic, mapped))
-                        .await?;
-                }
-                Err(error) => {
-                    debug!("Mapping error: {}", error);
+            debug!("Topic {:?}", message.topic.name);
+
+            // ? Validation for child device ID?
+            match get_child_id_from_topic(message.clone().topic.name) {
+                Some(child_id) if child_id.is_empty() => {
+                    let error = MapperError::InvalidChildId { id: child_id };
+                    error!("Child ID error: {}", error);
                     self.client
                         .publish(mqtt_client::Message::new(
                             &self.config.errors_topic,
                             error.to_string(),
                         ))
                         .await?;
                 }
+                Some(child_id) => {
+                    info!("Child ID {:?}", child_id);
+                    dbg!(&children);
+                    if !children.contains(&child_id) {
+                        children.insert(child_id.clone());
+                        if let Some(message) = self
+                            .converter
+                            .convert_child_device_creation(child_id.as_str())
+                        {
+                            self.client.publish(message).await?;
+                        };
+                    }
+
+                    match self
+                        .converter
+                        .convert_child_device_payload(message.payload_str()?, child_id.as_str())
+                    {
+                        Ok(mapped) => {
+                            self.client
+                                .publish(mqtt_client::Message::new(&self.config.out_topic, mapped))
+                                .await?;
+                        }
+                        Err(error) => {
+                            error!("Mapping error: {}", error);
+                            self.client
+                                .publish(mqtt_client::Message::new(
+                                    &self.config.errors_topic,
+                                    error.to_string(),
+                                ))
+                                .await?;
+                        }
+                    }
+                }
+                None => {
+                    // parent device
+                    match self.converter.convert(message.payload_str()?) {
+                        Ok(mapped) => {
+                            self.client
+                                .publish(mqtt_client::Message::new(&self.config.out_topic, mapped))
+                                .await?;
+                        }
+                        Err(error) => {
```

Comment:
This error handling is repeated for the 3 cases. Better to have the whole
`match get_child_id_from_topic(message.clone().topic.name) { ... }`
returning a `Result<Message, ConversionError>` and then publish the response on the appropriate topic.